File: back/script/cluster.py
import mysql.connector
import pandas as pd
from sklearn.cluster import KMeans
import json
import sys

# Récupérer le nombre de clusters à partir des arguments
num_clusters = int(sys.argv[1]) if len(sys.argv) > 1 else 3

# Connexion à la base de données
conn = mysql.connector.connect(
    host="localhost",
    user="root",
    password="",
    database="bddstquentin"
)

# Requête pour récupérer les données des arbres
query = "SELECT id_arbre, longitude, latitude FROM arbre"
df = pd.read_sql(query, conn)

# Fermer la connexion
conn.close()

# Clustering
kmeans = KMeans(n_clusters=num_clusters)
df['cluster'] = kmeans.fit_predict(df[['longitude', 'latitude']])

# Convertir les résultats en JSON
results = df.to_dict(orient='records')

# Écrire les résultats dans un fichier JSON
with open('./clusters.json', 'w') as f:
    json.dump(results, f)

File: back/php/cluster.php
<?php
header('Content-Type: application/json');
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: GET, POST, OPTIONS');
header('Access-Control-Allow-Headers: Content-Type, Access-Control-Allow-Headers, Authorization, X-Requested-With');

require_once 'dbconnect.php';
require_once 'data_encode.php';

$conn = connectDB();

$numClusters = isset($_GET['numClusters']) ? intval($_GET['numClusters']) : 2;
$pythonScriptPath = escapeshellarg(__DIR__ . '/../script/cluster.py');
$jsonFilePath = __DIR__ . '/../script/clusters.json';
$arbreJsonPath = __DIR__ . '/../script/arbres.json';

$pythonInterpreter = 'python';

// Requête pour récupérer les 200 premiers arbres
$sql = "SELECT id_arbre, latitude, longitude FROM arbre LIMIT 200";
$result = $conn->query($sql);

if ($result === false) {
    sendError(500);
}

if ($result->num_rows > 0) {
    $arbres = [];
    while ($row = $result->fetch_assoc()) {
        $arbres[] = $row;
    }

    // Écrire les données des arbres dans un fichier JSON
    if (file_put_contents($arbreJsonPath, json_encode($arbres)) === false) {
        sendError(500);
    }
} else {
    sendError(404);
}

$command = "$pythonInterpreter $pythonScriptPath $numClusters 2>&1";
$output = shell_exec($command);

if ($output === null) {
    sendError(500);
}

if (file_exists($jsonFilePath) && is_readable($jsonFilePath)) {
    $jsonData = file_get_contents($jsonFilePath);
    sendJsonData(['message' => 'Python script executed successfully', 'num' => $numClusters, 'command' => $command, 'data' => json_decode($jsonData)], 200);
} else {
    sendError(500);
}

$conn->close();
?>
